feat: Document dynamic sampling (#2954)
getsentry/sentry-docs: src/docs/product/data-management-settings/dynamic-sampling/index.mdx

---
title: Dynamic Sampling
sidebar_order: 10
description: "Learn about the different methods for controlling the amount and type of data sampled for your project."
---

<Note>

> This content is provided on a working branch as part of a project that is in progress. We anticipate it will be changed and modified as the project progresses.

Reviewer: `changed` and `modified` is not redundant?

Author: yes it is, will change

</Note>

Dynamic sampling addresses the need to easily change and manage what is sampled after an application is deployed. In addition, dynamic sampling enables more sophisticated sampling decisions that span multiple transactions.

When you select your sampling type, you'll find that dynamic sampling:

- prioritizes impactful data
- enables observation of complete traces
- maintains a representative sample
- requires minimal configuration and no redeployment

## Sampling Types

> Sentry offers the ability to sample in two different ways. You can sample [client-side in your SDK](/platform-redirect/?next=/configuration/sampling/) during SDK initialization or by [setting dynamic sampling rates](TODO: add link to new content) in Filters & Sampling for your Project in sentry.io. Learn more about our [sampling types and the benefits of each](/product/data-management-settings/dynamic-sampling/sampling-types/).

Reviewer: "in Filters & Sampling for your Project in sentry.io" should we mention settings here?

---
title: Dynamic Sampling
sidebar_order: 10
description: "Learn about the different methods for controlling the amount and type of data sampled for your project."
---

> <Alert level="info">
> 
> This content is provided on a working branch as part of a project that is in progress. We anticipate it will be changed as the project progresses.
> 
> </Alert>

Reviewer 2: ```suggestion
<Note>

This content is provided on a working branch as part of a project that is in progress. We anticipate it will be changed as the project progresses.

</Note>
```

When you select your sampling type, you'll find that dynamic sampling:

- helps you prioritize impactful data
- enables observation of complete traces
- maintains a representative sample
- requires minimal configuration and no redeployment

## Getting Started
Quickly get up to speed and see [how to use dynamic sampling](/product/data-management-settings/dynamic-sampling/getting-started/).

## Sampling Types

Sentry offers the ability to sample in two different ways:

1. [Client-side in your SDK](/platform-redirect/?next=/configuration/sampling/) during SDK initialization
2. [Dynamic sampling rates](./configuration/) in Filters & Sampling for your Project in sentry.io.

Learn more about our [sampling types and the benefits of each](/product/data-management-settings/dynamic-sampling/sampling-types/).

## Dynamic Sampling Configuration

Dynamic sampling operates on both error events and transaction events. When creating rules for transactions, you can
further determine whether to apply dynamic sampling rules to transactions independently or as part of a trace. Learn
more about [how the rules are applied](/product/data-management-settings/dynamic-sampling/configuration/).

> ## Example

Reviewer 2: ```suggestion
## Examples of Dynamic Sampling
```

Determining which type of dynamic sampling to set varies based on what you're trying to accomplish. Review our
[comprehensive examples](/product/data-management-settings/dynamic-sampling/example/) that illustrates when
different sampling methods might be useful.

> ## Miscellanea

Reviewer 2: ```suggestion
## Current Implementation Status
```

---
title: Dynamic Sampling
sidebar_order: 6
description: "Learn about the different methods for limiting the amount of data generated by your project."
---

TODO -  Should we use Dynamic Sampling or something like: "Centralized Sampling" or "Server Side Sampling" (I think
that, from a user's perspective, "Dynamic Sampling" doesn't convey much information).

# Why Dynamic Sampling

> Sentry SDKs are able to sample events by setting `traces_sample_rate` or `traces_sampler` during SDK initialisation

Reviewer 3: ```suggestion
You can set up sampling when you configure your SDKs by setting `traces_sample_rate` or `traces_sampler` during SDK initialization
```

> (TODO add link to .../PerformanceMonitoring/SamplingTransactions).

Reviewer 3: Note: I'll be moving this into the Configuration section for the SDKs to align with the Filtering content. But feel free to move forward. I'll take care of fixing any links I break in the process :)

> While this mechanism is very simple and works well in certain scenarios it has the downside of being difficult to
> change for scenarios where the application is distributed and redeployment is either not an option or very
> difficult (e.g. mobile applications).

Reviewer 3: ```suggestion
While this mechanism is simple and works well in many scenarios, it has the downside of being difficult to
change for scenarios where the application is distributed and redeployment is either not an option or very
difficult (for example, in mobile applications).
```

Dynamic Sampling addresses the need to easily change and manage what is sampled after an application is deployed,
as an additional benefit Dynamic Sampling enables making more sophisticated sampling decisions that span multiple
transactions.

> # Dynamic Sampling types

Reviewer 3: Use title case 
```suggestion
# Dynamic Sampling Types
```

> Dynamic sampling operates on two types of events: errors and transactions.

Reviewer 3: ```suggestion
Dynamic sampling operates two both error events and transaction events.
```

Author: Shouldn't we have a preposition there ?
Maybe:
Dynamic sampling operates on both error events and transaction events.

Reviewer 3: yes. Thanks, I dropped the preposition in my edit.

## Errors

> Error sampling decisions are taken based on the event data.

Reviewer 3: I think we mean that sampling decisions are _made_ based on event data? And can we provide an example, as this section is a bit confusing as it sounds as if it's almost random, which I doubt it is.

A sampling rule is selected from the available sampling rules by going through the list of dynamic sampling rules for
errors and picking up the first matching rule that matches the event data. Rule matching is described in more details
below.

## Transactions

Transactions can be considered either as independent events or in the context of all other transactions belonging to a

> trace (see TODO tracing link for more info on traces).

Reviewer 3: This link, perhaps? https://docs.sentry.io/product/performance/distributed-tracing/#what-is-tracing

> Sentry offers the possibility to sample transactions either independently or as part of a trace; there are different
> benefits for each case.

Reviewer 3: ```suggestion
With Sentry, you can either sample transactions independently or as part of a trace. Each option provides different
benefits.
```

> Sampling transactions using transaction traces is advantageous when trying to understand transactions in the context
> of the global system. In this case we are not interested in seeing a particular transaction, but the whole group of
> transactions in a trace so sampling should occur at the trace level.

Reviewer 3: ```suggestion
### Sampling Transactions as Part of a Trace

This approach is advantageous when trying to understand transactions in the context
of the global system. In this case, you are not interested in seeing a particular transaction. Rather, you would like to see the entire group of transactions in a trace. As a result, the best sampling option to sample at the trace level.
```

> When sampling transaction traces the rules sampling rules are based on the transaction context, this means that the

Reviewer 3: ```suggestion
When sampling transaction traces the sampling rules are based on the transaction context, which means that the
```

sampling decisions are based on the information extracted from the system that initiated the transaction. A

> transaction trace sample will rules are based on the following attributes (all attributes belong to the initial

Reviewer 3: I'm not sure what this is mean to say, but here's a suggested edit:
```suggestion
transaction trace sample rules are based on the following attributes (all attributes belong to the initial
```

transaction, the transaction that started the trace):

* project id
* release
* environment
* user segment

> 

Reviewer 3: An example would be very useful here.

> While sampling transaction traces gives a very good view of how transactions flow through the system, sometimes we are

Reviewer 3: Write in the second person ("you", not "we")
```suggestion
### Sampling Transactions Independently

While sampling transaction traces gives a very good view of how transactions flow through the system, sometimes you are
```

interested in analysing transactions from a particular service that happens not to be the initiator of transaction

> (e.g. a backend server). If we are interested in creating rules based on attributes specific to our service, for

Reviewer 3: Avoid Latin terms (e.g., i.e., and etc.)
```suggestion
(for example, a backend server). If you are interested in creating rules based on attributes specific to a service, for
```

> example we have a suspicion that a particular release might have some performance problems, we can sample based on
> transaction event.

Reviewer 3: ```suggestion
example you have a suspicion that a particular release might have some performance problems, you can sample based on
transaction event.
```

Individual transaction rules are based on the transaction attributes and not on the trace context attributes and
therefore can target individual services that are not initiating traces.

> 

Reviewer 3: An example would be very useful here.

(TODO In the future we can say that these rules can use more attributes).

> # Relation with inbound filters

Reviewer 3: ```suggestion
# Dynamic Sampling and Inbound Filters
```

(TODO Decide if we should keep this since we'll probably remove inbound filters in the future.)

> Inbound filters are a related but different concept. At a high level inbound filters may be viewed as a coarse sampling
> technique where only 0% and 100% sampling rates are available. The main difference between how filters and sampling
> rules operate is that filters compose while sampling rules do not, the example below will clarify how they operate.

Reviewer 3: ```suggestion
Inbound filters are a related concept, withe key differences from Dynamic Sampling. Inbound filters may be viewed as a coarse sampling technique, where only 0% and 100% sampling rates are available. The key difference between filters and sampling is that filters compose while sampling rules do not, the example provides an illustration:
```

> If we have two filters defined, one that removes all requests coming from Internet Explorer 9 or lower and one that
> filter requests coming from an application with release 1.* the resulting effect will be that all events that are
> coming either from Explorer 9 or that are coming from a client with release 1.* will be filtered out.

Reviewer 3: Might we explain that this is similar to a join statement? 
```suggestion
With two filters defined - one that removes all requests coming from Internet Explorer 9 or lower, and one that
filters requests coming from an application with release 1.x - means that all events 
coming either from Explorer 9 or coming from a client with release 1.x will be filtered out.
```

Author: No, a join statement would be the wrong concept  (a join is a cartesian product of two sets). I think the easiest way to clarify what happens here is a logical "or" of two conditions. If we want to use sets to explain the concept the operation is a "union" (non discriminated). In this case, for me, thinking in terms of a logical "or" of two conditions is a bit more intuitive than thinking about sets (which would be the union in the space of all possible events of the subset of events that have release '1.*' and the subset of events that have the user agent "Explorer 9" :-) ).

Author: Regarding 1.x instead of 1.*  I'm probably biased but I think that 1.* would be better. The reason is that this is how we specify `glob` patterns, this is actually how a user would specify the release in the rules. So specifying '1.3.*' would match both release 1.3.1 and 1.3.7 but will **not** match 1.2.1. We are using `glob` matching in various places so most users will probably know what we are talking about. For now I'll leave it with '*' instead of 'x' but if you still feel that 'x' is clearer I'll change it.

Reviewer 3: The "or" concept is better, I agree.

I don't have a strong opinion about using an asterisk rather than `1.x`, but asterisks are used for formatting bold (two asterisks) and italic (one asterisk) in Markdown, so you may end up with bolded or italicized statements where you don't expect them.

> If we have two sampling rules defined, first sampling on release 1.* at 10%, and the second sampling on "prod1"
> environment at 20% the rules will not compose in any meaningful way. A request coming for release 1.1 and prod1 will
> match the first rule and will be sampled at 10% , A request for release 1.1 & prod2 will still match the first rule and
> will be sampled at 10% while a request for release 2.1 and prod1 will match the second rule and will be sampled at 20%

Reviewer 3: I think the term "compose" here isn't quite what we want to say. Do we mean "combine"?
```suggestion
With sampling, if we have two sampling rules defined, first sampling on release 1.x at 10%, and the second sampling on `prod1` environment at 20%. the rules will not compose in any meaningful way. That is, a request coming for release 1.1 and `prod1` will match the first rule and will be sampled at 10%. A request for release 1.1 and `prod2` will still match the first rule and will be sampled at 10%. A request for release 2.1 and `prod1` will match the second rule and will be sampled at 20%.
```

Author: I did mean compose as in  "functional composition" (https://en.wikipedia.org/wiki/Function_composition) which is what effectively happens, but combine works equally well for me.

Reviewer 3: Great. Thanks for clarifying.
